### ISSUE TITLE
Corrupt JSON output while using withJson but json_encode works (Slim 3)

### ISSUE THREAD

**Issue by rkkoszewski** [slimphp/Slim#2278]
Hi, 

I get different results when I use:
```
return $response->withJson($array);
```
Than when I use:
```
$response->getBody()->write(json_encode($array));
return $response->withHeader('Content-Type', 'application/json;charset=utf-8');
```

When checking the result with my browser, there seems to be no visual difference between both responses (content and headers). But it seems that in the withJson there is something that corrupts the JSON when used together with a JSON parser library in Java that says throws as exception "IlegalStateException: Expected BEGIN_OBJECT but was STRING", which basically means there is some kind of garbage at the begining of the JSON payload. This does not happen when the second example is used, and everything gets parsed correctly.

Diving into the withJson method I found what is causing the issue.

First of all ..
```
$response = $this->withBody(new Body(fopen('php://temp', 'r+')));
```
.. seems to be causing the issue. By removing this part the problem is almost fixed.

Secondly this part  ..
```
$response->body->write($json = json_encode($data, $encodingOptions));
```
.. has to be changed to ..
```
$response->getBody()->write($json = json_encode($data, $encodingOptions));
```
in order to get a clean and valid JSON response.
(body has been replaced with getBody())
[when the "..new Body.." part is present, corrupt JSON payload is generated no matter if body or getBody is used]

Like this there seems to be no more JSON corruption. But I don't understand why. What is the "new Body(fopen('php://temp', 'r+'))" part doing? Why is body instead of getBody used and why does it yield different results? Cold it be some encoding issues when the php://temp is created? Is there any fix for it?

Is this a bug or is it supposed to be like this?

Some background on what I'm trying to achieve:
I'm building a REST application that communicates with a Java application by using JSON. It seems tough that the JSON library that is used in the Java application does not like spaces or weird characters on the beginning of the response.

I hope somebody can give me some input on this. As workaround I can just use "return json_encode(.." but if the issue can be fixed to make it work with the official asJson method, I believe that would be the best option.

Greetings and thanks.

**Comment by akrabat** [comment]
Can you create a one php file example app with two routes in it where one works and the other doesn't please?

**Comment by geggleto** [comment]
@akrabat I did notice that there wasn't a specific unit test covering withJson ...

**Comment by RyanNerd** [comment]
@geggleto Does [this test](https://github.com/slimphp/Slim/blob/3.x/tests/Http/ResponseTest.php#L286) provide adequate coverage?

**Comment by geggleto** [comment]
@RyanNerd I'll take a look at it

**Comment by rkkoszewski** [comment]
Hi, so I've found what was causing the issue. The method that was giving me the error had a complex POST request, and therefore in order to test it manually I used an other method that can be accessed by a simple GET request, so I could debug the result. The reason why both ways were showing for me the same result is because it was the same. I was in belief that I was using the same code snippet, and I didn't notice that the POST method I was using the $request instead of the $response to call the withJson method. Both $request and $response look visually similar and that's why I didn't notice that I was using the wrong variable. So it was throwing an exception actually, but I've never seen it because I was testing it on the GET method.

I would like to thank everybody who commented and tried to help me solve the issue. At the end it was just my fault. Everything seems to be working perfectly fine now.

Greetings and thanks.
